### PR TITLE
[Freestanding] Only run tasks when they are awaited up on in task-to-thread model

### PR DESCRIPTION
**Pull request by rokhinip**
In the task-to-thread model, there are no threading mechanisms by which work could be offloaded onto another thread. When creating structured concurrency tasks however, we know that they will be awaited upon by the calling parent task which can therefore execute the child task at the point of await.

Rdar: rdar://92347604

### CODE REVIEW

**Comment by rokhinip** [comment]
@swift-ci please test

**Comment by rokhinip** [comment]
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset

File: stdlib/public/Concurrency/AsyncLet.cpp [hunk 1]
```diff
@@ -167,7 +167,13 @@ void swift::swift_asyncLet_start(AsyncLet *alet,
                                  void *closureEntryPoint,
                                  HeapObject *closureContext) {
   auto flags = TaskCreateFlags();
+#if SWIFT_CONCURRENCY_TASK_TO_THREAD_MODEL
```
Comment (ktoso) [inline stdlib/public/Concurrency/AsyncLet.cpp:170]: I find the name of this very confusing; How is this "... to thread", it sounds like we always execute in parent thread, or in other words "task on calling thread" or something...?

File: stdlib/public/Concurrency/Task.cpp [hunk 2]
```diff
@@ -574,12 +603,16 @@ static inline bool isUnspecified(JobPriority priority) {
   return priority == JobPriority::Unspecified;
 }
 
-static inline bool taskIsUnstructured(JobFlags jobFlags) {
-  return !jobFlags.task_isAsyncLetTask() && !jobFlags.task_isGroupChildTask();
+static inline bool taskIsStructuredConcurrency(JobFlags jobFlags) {
```
Comment (ktoso) [inline stdlib/public/Concurrency/Task.cpp:606]: This reads weird, `taskIsStructured` would be the the usual wording and pair up well with`taskIsUnstructured`

File: stdlib/public/Concurrency/TaskGroup.cpp [hunk 3]
```diff
@@ -284,15 +284,24 @@ class TaskGroupImpl: public TaskGroupTaskStatusRecord {
...
+  // model, only the parent task which created the task group can
+  //
+  //   (a) add child tasks to a group
+  //   (b) run the child tasks
+  //
+  // So we shouldn't need to worry about coordinating between child tasks and
+  // parents in a task group
+  void lock() const {}
+  void unlock() const {}
+#else
```
Comment (ktoso) [inline stdlib/public/Concurrency/TaskGroup.cpp:299]: 👍

**Review by ktoso — APPROVED** [review]
Looks okey AFAICS, though the naming of this mode is very weird sounding, any better ideas? The "task to thread" doesn't really mean anything when one reads it hm. "to" what, it is not that a task is mapped "to" a thread after all, so it sounds even misleading somewhat. Is it perhaps "run tasks inline" mode, that's what it does after all 🤔

File: stdlib/public/Concurrency/Task.cpp [hunk 4]
```diff
@@ -187,6 +187,29 @@ FutureFragment::Status AsyncTask::waitFuture(AsyncTask *waitingTask,
...
+    // executor
+    swift_job_run(this, ExecutorRef::generic());
+
+    SWIFT_TASK_DEBUG_LOG("[RunInline] Switching back from running %p to now running %p", this, oldTask);
+
+    // We now are back in the context of the waiting task and need to reevaluate
+    // our state
+    _swift_task_setCurrent(oldTask);
+    queueHead = fragment->waitQueue.load(std::memory_order_acquire);
+    continue;
```
Comment (ktoso) [inline stdlib/public/Concurrency/Task.cpp:211]: Ok makes sense 👍

File: stdlib/public/Concurrency/Task.cpp [hunk 5]
```diff
@@ -187,6 +187,29 @@ FutureFragment::Status AsyncTask::waitFuture(AsyncTask *waitingTask,
       escalatedPriority = waitingStatus.getStoredPriority();
     }
 
+#if SWIFT_CONCURRENCY_TASK_TO_THREAD_MODEL
+    // In the task to thread model, we will execute the task that we waiting on,
```
Comment (mikeash) [inline stdlib/public/Concurrency/Task.cpp:191]: Missing "are" in "we waiting on."

File: stdlib/public/Concurrency/TaskGroup.cpp [hunk 6]
```diff
@@ -694,7 +730,7 @@ static void swift_taskGroup_wait_next_throwingImpl(
     OpaqueValue *resultPointer, SWIFT_ASYNC_CONTEXT AsyncContext *callerContext,
     TaskGroup *_group,
     ThrowingTaskFutureWaitContinuationFunction *resumeFunction,
-    AsyncContext *rawContext) {
+    AsyncContext *rawContext) SWIFT_OPTNONE {
```
Comment (mikeash) [inline stdlib/public/Concurrency/TaskGroup.cpp:733]: Is `SWIFT_OPTNONE` supposed to be here? If so, can we have a comment explaining it?

File: stdlib/public/Concurrency/TaskGroup.cpp [hunk 7]
```diff
@@ -739,16 +775,25 @@ static void swift_taskGroup_wait_next_throwingImpl(
   }
 }
 
-PollResult TaskGroupImpl::poll(AsyncTask *waitingTask) {
+PollResult TaskGroupImpl::poll(AsyncTask *waitingTask) SWIFT_OPTNONE {
```
Comment (mikeash) [inline stdlib/public/Concurrency/TaskGroup.cpp:778]: Likewise for this `SWIFT_OPTNONE`.

File: stdlib/public/Concurrency/TaskPrivate.h [hunk 8]
```diff
@@ -798,7 +803,7 @@ inline void AsyncTask::flagAsSuspended() {
 // READ ME: This is not a dead function! Do not remove it! This is a function
 // that can be used when debugging locally to instrument when a task actually is
 // dealloced.
-inline void AsyncTask::flagAsCompleted() {
+inline void AsyncTask::flagAsCompleted() SWIFT_OPTNONE {
```
Comment (mikeash) [inline stdlib/public/Concurrency/TaskPrivate.h:806]: And again for `SWIFT_OPTNONE`.

**Review by mikeash — COMMENTED** [review]
Looks good overall. A few little things. I agree with @ktoso that the name is confusing. Not sure what would be better though.

File: stdlib/public/Concurrency/AsyncLet.cpp (hunk 1 above)
Comment (rokhinip) [reply stdlib/public/Concurrency/AsyncLet.cpp:170]: Yea English is hard 

But this task-to-thread terminology was already present and introduced by @nate-chandler so I am simply extending its use in this PR. If we come up with a better name, we can make a separate change to adjust it across the board.

File: stdlib/public/Concurrency/Task.cpp (hunk 2 above)
Comment (rokhinip) [reply stdlib/public/Concurrency/Task.cpp:606]: Good point. Will fix

File: stdlib/public/Concurrency/TaskGroup.cpp (hunk 6 above)
Comment (rokhinip) [reply stdlib/public/Concurrency/TaskGroup.cpp:733]: Nope, I did it for easier debugging.

Comment (mikeash) [reply stdlib/public/Concurrency/TaskGroup.cpp:733]: I'll have to remember that trick.

**Comment by rokhinip** [comment]
@swift-ci please test

**Comment by rokhinip** [comment]
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset

**Review by mikeash — APPROVED** [review]

**Pull request merged**